File: Python/Count Binary Substrings.py
class Solution:
    """
    Time:   O(n)
    Memory: O(1)
    """

    def countBinarySubstrings(self, s: str) -> int:
        substrings = 0
        prev, curr = 0, 1

        for i in range(1, len(s)):
            if s[i - 1] != s[i]:
                substrings += min(prev, curr)
                prev, curr = curr, 1
            else:
                curr += 1

        return substrings + min(prev, curr)

File: Java/Count Binary Substrings.java
public class Solution {

    /**
     * Time:   O(n)
     * Memory: O(1)
     */
    public int countBinarySubstrings(String s) {
        int substrings = 0;
        int prev = 0, curr = 1;

        for (int i = 1; i < s.length(); ++i) {
            if (s.charAt(i - 1) != s.charAt(i)) {
                substrings += Math.min(prev, curr);
                prev = curr;
                curr = 1;
            } else {
                ++curr;
            }
        }

        return substrings + Math.min(prev, curr);
    }
}
